### ISSUE TITLE
pcscd: Error Reader Exclusive

### ISSUE THREAD

**Issue by chrisaq** [drduh/YubiKey-Guide#277]
When having pcscd and scdaemon running at the same time (in linux), I regularly get "accesss sharing violation", "Error Reader Exclusive", "ccid_usb.c:657:OpenUSBByName() Can't claim interface 1/4: LIBUSB_ERROR_BUSY" and similar errors in the pcscd logs.

When getting these errors the yubikey is unavailable to pcscd, restarting pcscd will temporarily make it available again.

This is caused by, as far as I've been able to figure out, scdaemon claiming exclusive access to the usb device. The scdaemon.conf "shared-access" or "card-timeout 1" does not solve the problem for me.

Maybe this issue could be explored in the document and a possible solution be added (if known).

**Comment by donnex** [comment]
I'm not sure if this is exactly the same issue but it could be worth a try. I previously had issues when using both oath (totp) and gpg at the same time. My YubiKey wasn't responding when trying to use both applications in a short time frame.

After some digging around I found https://ludovicrousseau.blogspot.com/2019/06/gnupg-and-pcsc-conflicts.html

I added `disable-ccid` to my `scdaemon.conf`.

**Comment by chrisaq** [comment]
> I'm not sure if this is exactly the same issue but it could be worth a try. I previously had issues when using both oath (totp) and gpg at the same time. My YubiKey wasn't responding when trying to use both applications in a short time frame.
> 
> After some digging around I found https://ludovicrousseau.blogspot.com/2019/06/gnupg-and-pcsc-conflicts.html
> 
> I added `disable-ccid` to my `scdaemon.conf`.

Thank you for responding!

I do actually have `disable-ccid` in my scdaemon.conf, but that doesn't seem to make a difference for me.

It seems like ỳkman`is the only application I use that needs the pcscd interface.
Unfortunately I also use yubikey otp, which uses ykman/pcsc...

So far, while trying to figure this out, I stumbled across these options, but I have yet to find a combination that works:

```
card-timeout 1
pcsc-driver /usr/lib/libpcsclite.so
disable-ccid
shared-access
reader-port Yubico Yubi
```

**Comment by chrisaq** [comment]
So I have yet to find a real solution to this using standard software distributions.

However, for those who use Arch, there is a patched version of gnupg/scdaemon that allows shared access to the yubikey, it is available here:

https://aur.archlinux.org/packages/gnupg-scdaemon-shared-access/

It allows for the `pcsc-shared` config option in `scdaemon.conf` which allows other processes (pcscd) to use the yubikey as well.

**Comment by chrisaq** [comment]
A short update for those having the same problem.

Even the above package for Arch does not solve the problem completely for me.

However, the lock is not as hard to get around, restarting pcscd is no longer needed, and there is a new error message for `gpg --card-status`:

`gpg: OpenPGP card not available: Card error`

Now running any `ykman` command, and waiting a few seconds, lets gpg/scdaemon have access again, and scdaemon never seems not lock out pcscd.

I have yet to figure out what causes gpg/scdaemon to lose access, but I suspect it might be another process, like a browser or similar.

Will keep updating this ticket until a solution is found.

**Comment by chrisaq** [comment]
The latest finding is that the yubikey gpg/scd works without issues until a browser is opened:

```
Arch linux
firefox 90.0.2-1
gnupg 2.2.29-1
pcsclite 1.9.3-1
```

On a different machine I have no issues at all, even with browser.

```
Arch linux
firefox 89.0.1-1
gnupg 2.2.28-2
pcsclite 1.9.1-1
```

On the problematic machine I run `watch gpg --card-status`, and it keeps running without error for hours until a browser is opened. When the browser is opened, it errors out within seconds. If I run any `ykman` command it works again for a few seconds, then locks up.

I will try to switch software versions around on the broken machine to see if that might reveal the real source of the problem.

**Comment by stibi** [comment]
Thanks for you research and notes here. This is bugging me a lot, too. `pass` and `ykman oath` is huge part of my daily workflow, so when it locks up, it's very annoying.

**Comment by stibi** [comment]
Btw: https://wiki.archlinux.org/title/GnuPG#Shared_access_with_pcscd

> Starting from version 2.2.28 LTS and 2.3.0 you can enable shared access by modifying your scdaemon.conf file and adding pcsc-shared line end of it. 

But it's not fully clear to me what is the difference between `shared-access` and `pcsc-shared` and if the patched [`gnupg-scdaemon`](https://aur.archlinux.org/packages/gnupg-scdaemon-shared-access/) is actually needed or not.

**Comment by stibi** [comment]
Oh I see now. Both are doing the same probably.

The `shared-access` config option makes only sense with the `gnupg-scdaemon-shared-access` patched Archlinux package, it connects to pcsc using the `PCSC_SHARE_SHARED` flag. (https://aur.archlinux.org/cgit/aur.git/tree/scdaemon_shared-access.patch?h=gnupg-scdaemon-shared-access)

And the `pcsc-shared` does the same, but for the upstream, not patched gnupg scdaemon, starting from version 2.2.28 LTS and 2.3.0.

- https://github.com/gpg/gnupg/blob/25ae80b8eb6e9011049d76440ad7d250c1d02f7c/NEWS#L111
- https://github.com/gpg/gnupg/blob/25ae80b8eb6e9011049d76440ad7d250c1d02f7c/doc/scdaemon.texi#L229
- https://github.com/gpg/gnupg/blob/25ae80b8eb6e9011049d76440ad7d250c1d02f7c/scd/scdaemon.c#L154

No big reveal here ... just I can revert back to the upstream `gnupg` arch package and go with the `pcsc-shared` instead of `shared-access` in `~/.gnupg/scdaemon.conf`. That's nice :)

**Comment by klausenbusk** [comment]
Do you have `opensc` installed? Firefox uses `/usr/lib/opensc-pkcs11.so` provided by `opensc` for accessing the smartcard. IIRC I removed `opensc` and it solved the issue.

**Comment by chrisaq** [comment]
@stibi Yes, you are correct about the `shared-access` and `pcsc-shared` stuff.
There are so many moving parts and I haven't been good enough to describe all steps.

In fact I was hoping it would just be a few posts and a permanent solution would be found, but it hasn't worked out that way, still having issues.

Currently can't use `ykman oath` and gpg together, I have to shut down pcscd or it fails, and get these error messages from scdaemon:
`pcsc_connect failed: sharing violation`

Some useful commands for other people playing around with this:

`ykman oath accounts list`
`gpgconf --kill scdaemon`
`sudo journalctl /usr/lib/gnupg/scdaemon`

```
# /etc/systemd/system/pcscd.service.d/override.conf
[Service]
ExecStart=
ExecStart=/usr/bin/pcscd --debug --foreground --auto-exit $PCSCD_ARGS

```

**Comment by chrisaq** [comment]
@klausenbusk Thanks for responding!
I have uninstalled it now. Perhaps the browsers keep polling the pcscd so much that it blocks out other connections? Worth a try.

**Comment by donnex** [comment]
Let's hope that resolves your problem. I can't figure out how my setup differs from yours other than this package.

I'm running Arch with the latest GPG package from pacman. Currently using GPG, oath/totp and U2F with my YubiKey 5C. 

Before adding `disable-ccid` I had lockups where I had to remove and reinsert my key whenever I was accessing both oath and GPG in a short time span. With `disable-ccid` added I can use both of them with the only caveat that I have to reenter my pin when switching from oath back to GPG.

**Comment by chrisaq** [comment]
> I'm running Arch with the latest GPG package from pacman. Currently using GPG, oath/totp and U2F with my YubiKey 5C.
> 
> Before adding `disable-ccid` I had lockups where I had to remove and reinsert my key whenever I was accessing both oath and GPG in a short time span. With `disable-ccid` added I can use both of them with the only caveat that I have to reenter my pin when switching from oath back to GPG.

Thanks for the trying to help out with this!
Since you use it like I do and use the same distro, I gave it another try, and I might have found something interesting.

What I found strange was that in the start of this thread it was scdaemon that got an exclusive lock on the yubikey because of the missing `pcre-shared` option. In the latest posts the problem is reversed, scdaemon can't get access and complains about sharing violation.

So I looked over the config file, and removed `pcsc-driver /usr/lib/libpcsclite.so`, and then both worked for a few minutes until pcscd was locked out, so it  switched around. Then I restarted both while reactivating the pcsc-driver option, and now both have worked for about 10 minutes. 

Let's just say I'm a bit confused!

Anyways, this is my current config which has worked for 10 minutes on Arch:

```
card-timeout 1
pcsc-driver /usr/lib/libpcsclite.so
disable-ccid
pcsc-shared
log-file /home/chrisq/scdaemon.log
debug-level advanced
```

pcsc:
`pcscd --debug --foreground --auto-exit`

Can you post your complete `scdaemon.conf`?

**Comment by donnex** [comment]
Any luck with this issue?

I've recently reinstalled/replaced one of my machines and got this working as before. One interesting thing I found out was that my new laptop had a built in card reader. Initially this reader caused issues with pcscd trying to use it instead of my YubiKey. I guess it was possible to configure this but I just disabled the card reader and everything worked great.

I'm running with the following installed packages:
 
- ccid
- gcr
- gnupg
- hopenpgp-tools
- pcsclite

I don't have anything configured in my `scdaemon.conf` other than `disable-ccid`.

**Comment by chrisaq** [comment]
> Any luck with this issue?
> 
> I've recently reinstalled/replaced one of my machines and got this working as before. One interesting thing I found out was that my new laptop had a built in card reader. Initially this reader caused issues with pcscd trying to use it instead of my YubiKey. I guess it was possible to configure this but I just disabled the card reader and everything worked great.
> 
> I'm running with the following installed packages:
> 
>     * ccid
> 
>     * gcr
> 
>     * gnupg
> 
>     * hopenpgp-tools
> 
>     * pcsclite
> 
> 
> I don't have anything configured in my `scdaemon.conf` other than `disable-ccid`.

@donnex 

Still do not have a perfect solution, but it's at least in a state where I can use it.
Before starting a work session I do a `gpgconf --kill scdaemon && systemctl restart pcscd`.
If I do not do that, `ykman` can't find the yubikey, no obvious reason for that in the pcscd-debug log.

After the kill + restart, I can use both `ykman` and `gpg`, except for a few seconds after running `ykman`, which I think is caused by switching the slot/applet in the yubikey.

How I test this is running `watch gpg --card-status` and `watch -n 10 ykman oath accounts list` in separate terminals and watching the output. So far it doesn't seem to hang, but on the other hand, I have no idea what makes it hang and require the kill+restart in the first place.

Thank you for the config+packages information and the card reader tips, something new to try!

I'll keep updating this ticket even as I got a partially working solution since the unknown reason for it partially failing is annoying to me.

**Comment by donnex** [comment]
Yeah, really annoying :/

I tried running the watch commands and had no problems with that. Other than some of the times the `gpg --card-status` gave an error because it couldn't access the YubiKey. Everything recovered and I'm able to use both of the apps without any restart of pcscd.

What does your `pcscd.service` logs say?

**Comment by chrisaq** [comment]
> Yeah, really annoying :/
> 
> I tried running the watch commands and had no problems with that. Other than some of the times the `gpg --card-status` gave an error because it couldn't access the YubiKey. Everything recovered and I'm able to use both of the apps without any restart of pcscd.
> 
> What does your `pcscd.service` logs say?

I set pcscd to debug logging, and it's not that readable really, but I do get the occasional exclusive access error there.

Wow, now, after removing the card-timeout, pcsc-shared and pcsc-driver from the scdaemon.conf it seems to work just as good as with them, even though those option supposedly fixes some of the issues regarding exclusive access and so on. It's been so much back and forth though, that I'm not taking it for granted that it's actually working, what happened to the exclusive access errors from pcscd from before pcsc-shared option?

Anyways, a good idea to cut all config to a minimum like you have and take it from there.

The card-reader issue you were having might be interesting too, as I have a total of 3 card readers on monitors, docking station and laptop, do you remember what kind of error you were getting and in what service?

I fear the lock ups I still get before kill+restart might be caused by hardware like with your card-reader, but perhaps the thunderbolt port the yubikey is connected to, which is why it seems to random and not easily reproducible. That would also explain why a different usb-a yubikey4 it is working perfectly on a different laptop with the same config.

**Comment by donnex** [comment]
Interesting, let's hope you'll get this sorted out :)

I got `10230833 winscard.c:286:SCardConnect() Error Reader Exclusive` errors when running the watch commands. Other than that no errors and almost no logging occurs at all for me. Only some `LIBUSB_ERROR_NO_DEVICE` when I remove my key.

At the moment I can't check or remember what errors I got when my card reader was enabled. I got random issues where no card could be found and had to restart `pcscd`. I think the logs said something about using the first card (not the YubiKey but the card reader). Maybe try to disable debug logging and check what yours says?

When researching that issue I found the following command that outputs all devices. Worth a try to see if your other readers shows up.

```
echo scd getinfo reader_list | gpg-connect-agent --decode | awk '/^D/ {print $2}'
``` 

If you have any other devices than just the YubiKey I would look into configuring scdaemon or maybe pcscd (not sure) to only use a specific card. I guess that should be possible. If you're not able to disable the other readers.

**Comment by chrisaq** [comment]
@donnex Thanks for all your input, much appreciated!

Now I've had my yubikey lock up even without scdaemon running at all, only pcscd, and it was fixed by restarting pcscd, which sort of sounds like your situation with the card readers.
When keeping the `watch ykman` command runnning it didn't seem to lock up, but after exiting it and using the machine normally, I noticed it had locked up again maybe 30 minutes later.

So is the unavailability of the yubikey caused by pcscd getting confused by other card readers, are other applications, like a web browser attaching through pcscd and not releasing again, or perhaps a totally different process is attaching to the yubikey directly and causing the LIBUSB_ERROR_BUSY?
Still a mystery to me, and it would be so much easier if I knew how to trigger the problem.

Errors in the pcscd log, unknown cause and severity:
```

ccid_usb.c:641:OpenUSBByName() Found interface 2 but expecting 1
ccid_usb.c:643:OpenUSBByName() Wrong interface for USB device 1/6. Checking next one.
...
readerfactory.c:1120:RFInitializeReader() Open Port 0x200000 Failed (usb:1050/0407:libudev:1:/dev/bus/usb/001/006)
readerfactory.c:380:RFAddReader() Yubico YubiKey OTP+FIDO+CCID init failed.
...
ccid_usb.c:657:OpenUSBByName() Can't claim interface 1/6: LIBUSB_ERROR_BUSY
...


```

**Comment by donnex** [comment]
What a mystery :)

I can see that you've got OTP activated on the YubiKey. I don't use that application and have it disabled. No idea if this affects pcscd or not.

`Enabled USB interfaces: FIDO, CCID`

Did you check the list of readers with the command above?

**Comment by TheAifam5** [comment]
I have solution for that.

On Gentoo just edit `/lib/udev/rules.d/99-pcscd-hotplug.rules`:
```diff
# Copyright 1999-2019 Gentoo Authors
# Distributed under the terms of the GNU General Public License v2

# We add this here so that it runs after ccid's and ifd-gempc's rules;
# if we just added a pcscd-owned device, we hotplug the pcscd service.
-ACTION=="add", ENV{PCSCD}=="1", GROUP="pcscd", TAG+="systemd", ENV{SYSTEMD_WANTS}+="pcscd.service", RUN+="pcscd.sh"
+ACTION=="add|remove", ENV{PCSCD}=="1", GROUP="pcscd", TAG+="systemd", ENV{SYSTEMD_WANTS}+="pcscd.service", RUN+="pcscd.sh"
```

You can create a new one under `/etc/udev/rules.d/99-pcscd-hotunplug.rules` and paste this:
```diff
ACTION=="add|remove", ENV{PCSCD}=="1", GROUP="pcscd", TAG+="systemd", ENV{SYSTEMD_WANTS}+="pcscd.service", RUN+="pcscd.sh"
```

This will ensure that after removal, the pcscd daemon will never spam your logs trying to communicate with disconnected device.

Also `scdaemon` config looks like this: ~/.gnupg/scdaemon.conf
```
pcsc-driver /usr/lib64/libpcsclite.so
card-timeout 5
disable-ccid
pcsc-shared
```

It works after disconnecting, querying from `ykman` and `gpg --card-status`.

If you use `ykman list` and after that `gpg --card-status`, you will receive:
`gpg: OpenPGP card not available: General error` - just run the command again.

I think that `disable-ccid` can be removed but I am gonna leave as it.

**You see any improvements? Let me know! :)**

**Comment by drduh** [comment]
Closing this out, but please send PRs with any helpful troubleshooting steps or links others can refer to.

**Comment by dcunited001** [comment]
@rahra It could just be that one application attempts to access the hardware without going through the proper interfaces  

![Image](https://github.com/user-attachments/assets/c8782d15-8110-49ca-bacb-611db018f568)

this image may help a bit. i can't remember where I found it

**Comment by gsauthof** [comment]
@rahra [wrote](https://github.com/drduh/YubiKey-Guide/issues/277#issuecomment-2470039722)

> I know that this is an old thread but it seems that this problem still persist today. I observed this very same issue on a recent Ubuntu 24.04 with the latest Firefox 132.0.1. My conclusion of all of that above is that

FWIW, I observed similar issues on Fedora 41 and previous versions.

> 1. `disable-ccid` is a necessity.

For my uses, also `pcsc-shared` is required in `~/.gnupg/scdaemon.conf`:

```
disable-ccid
pcsc-shared
```
And `disable-application piv` to work-around PIN caching issues.

> 2. All other options do not have any conclusive effect but I found that it is Firefox messing up everything. Without Firefox everthing works and if there's some strange behavior, restart Firefox to fix it (I know, that's not a very satisfying solution...).

The `pcsc-shared` option should help with any Firefox interference.

In case you don't have any use case for the yk piv feature set in firefox you can also disable it completely in firefox.

See also my UNIX.se answer for details on how to do that and some references:

- [Browser polkitd pkla-check-authorization spam](https://unix.stackexchange.com/a/800181/1131)

NB: In case Firefox isn't delivered via a normal package on Ubuntu, but via a Snap, a different configuration may be necessary to disable piv integeration.
